core/src/main/java/org/infinispan/remoting/inboundhandler/BaseBlockingRunnable.java

```diff
@@ -28,10 +28,18 @@ protected BaseBlockingRunnable(BasePerCacheInboundInvocationHandler handler, Cac
 
    @Override
    public void run() {
+      InvocationStatus invocationStatus = null;
       try {
-         response = beforeInvoke();
-         if (response == null) {
-            response = handler.invokePerform(command);
+         invocationStatus = beforeInvoke();
+         switch (invocationStatus) {
+            case CONTINUE:
+               response = handler.invokePerform(command);
+               break;
+            case CACHE_NOT_FOUND:
+               response = CacheNotFoundResponse.INSTANCE;
+               break;
+            case DEFERRED:
+               return;
```

Comment:
> If releasing the lock immediately and using ordered messages is more performant, you could already adopt that for DIST_SYNC and use the RPC ack only to notify the originator about completion. 

Pedro's triangle work sends the ACK from the backup directly to the originator, but the primary also waits for an ACK before releasing the lock. I'm still worried about the edge cases, though :)

> I personally think that too much contention should not happen in practice, so optimizing that is low-priority.

I tend to agree, but contention might be more of a problem with the counters that Pedro is also working on. I expect a small set of counters, and a lot of writes to them.

> On the other hand, with per-sender ordering I am worried about global hiccups when a message is lost/dropped.

I'm not *too* worried. When a node receives a message out of order, it immediately sends a XMIT request to the sender, so the delay shouldn't be much more than 1 RTT.  When there's lots of allocation (or even blocking) during command processing, the receiver might take a long time to catch up, but I don't think that's a problem ATM.